### PR TITLE
nmstatectl: allow route rule filtering by interface

### PR DESCRIPTION
**Pull request by SudhanAruna**
Signed-off-by: SudhanAruna <sudhan.aruna.97@gmail.com>

Added Route Rule filter 
For Issue #935

### CODE REVIEW

File: nmstatectl/nmstatectl.py [hunk 1]
```diff
@@ -394,3 +396,15 @@ def _filter_routes(state, patterns):
...
     return routes
+
+
+def _filter_route_rule(state, patterns):
+
+    route_rules = {RouteRule.CONFIG: []}
+
+    for rule in state.get(RouteRule.CONFIG, []):
+        for pattern in patterns:
+            if fnmatch.fnmatch(rule, pattern):
```
Comment (cathay4t) [inline nmstatectl/nmstatectl.py]: A rule is pointing to a route table, route table contains route next hop to interface.

`fnmatch.fnmatch(rule, pattern)` will always return False.

Comment (SudhanAruna) [reply nmstatectl/nmstatectl.py]: @cathay4t 
I'll make the changes. I couldn't find example related or including route rule. Does the route table contains only next hop interface?

**Comment by SudhanAruna** [comment]
> 
> 
> The use of `fnmatch.fnmatch(rule, pattern)` is incorrect.
@cathay4t
@ffmancera  
I'll make the changes. I couldn't find example related or including route rule. In schema RouteRule does have next-hop-interface attribute (shall added it)?

File: libnmstate/schema.py [hunk 2]
```diff
@@ -72,6 +72,7 @@ class RouteRule:
     IP_TO = "ip-to"
     PRIORITY = "priority"
     ROUTE_TABLE = "route-table"
+    NEXT_HOP_INTERFACE = "next-hop-address"
```
Comment (ffmancera) [inline libnmstate/schema.py]: This is not needed. You can find the `next-hop-interface` in Route class. Thanks!

File: nmstatectl/nmstatectl.py [hunk 3]
```diff
@@ -394,3 +396,15 @@ def _filter_routes(state, patterns):
...
     return routes
+
+
+def _filter_route_rule(state, patterns):
+
+    route_rules = {RouteRule.CONFIG: []}
+
+    for rule in state.get(RouteRule.CONFIG, {}).get(RouteRule.ROUTE_TABLE, {}):
+        for pattern in patterns:
+            if fnmatch.fnmatch(rule[RouteRule.NEXT_HOP_INTERFACE], pattern):
```
Comment (ffmancera) [inline nmstatectl/nmstatectl.py]: Please, use the existing constant from the schema `Rule.NEXT_HOP_INTERFACE`.

File: tox.ini [hunk 4]
```diff
@@ -1,7 +1,6 @@
 [tox]
 envlist = black, flake8, pylint, py36, py38
 skip_missing_interpreters = True
-
```
Comment (ffmancera) [inline tox.ini]: This change is not needed.

**Comment by SudhanAruna** [comment]
> 
> 
> Thanks! Some comments below. Please, adjust the commit message. We only need only one `Signed-off-by` and a description of the feature. Thanks!

@ffmancera 
I have made the adjustments you mentioned and thank you for the comments.

**Comment by SudhanAruna** [comment]
> 
> 
> The code looks good but the commit message doesn't. Please consider this a commit title and description:
> 
> ```
> nmstatectl: allow route rule filtering by interface
> 
> "nmstatectl show <interface_name>" is now filtering the route rules by interface name.
> ```
> 
> Thanks!

Okey, I'll amend it. Thanks

**Comment by ffmancera** [comment]
Please, change the commit properly. You need to do `git commit --amend` and do the proper changes. Thanks.

**Comment by SudhanAruna** [comment]
> 
> 
> Please, change the commit properly. You need to do `git commit --amend` and do the proper changes. Thanks.

@ffmancera 
Sorry for that.

File: nmstatectl/nmstatectl.py [hunk 5]
```diff
@@ -394,3 +396,15 @@ def _filter_routes(state, patterns):
...
     return routes
+
+
+def _filter_route_rule(state, patterns):
+
+    route_rules = {RouteRule.CONFIG: []}
+
+    for rule in state.get(RouteRule.CONFIG, {}).get(RouteRule.ROUTE_TABLE, {}):
+        for pattern in patterns:
+            if fnmatch.fnmatch(rule[Route.NEXT_HOP_INTERFACE], pattern):
```
Comment (cathay4t) [inline nmstatectl/nmstatectl.py]: The rule rule does not have `Route.NEXT_HOP_INTERFACE`.

A suggestion on the approach. The `_filter_routes()` has got all interested routes, you just need to search whether
any rule is pointing to the route table of those routes.

Comment (SudhanAruna) [reply nmstatectl/nmstatectl.py]: @cathay4t 
I have made the changes so that route rule filters rule which match the table id in filtered route.

**Comment by SudhanAruna** [comment]
@cathay4t 
@ffmancera 
There is no example for route rule. Shall I work on that?

File: nmstatectl/nmstatectl.py [hunk 6]
```diff
@@ -394,3 +396,17 @@ def _filter_routes(state, patterns):
...
     return routes
+
+
+def _filter_route_rule(state):
+    """
+    return the rules for state's route rule that match the table_id of the
+    filtered route of state by interface
+    """
+    route_rules = {RouteRule.CONFIG: []}
+    route_table_id = state.get(Route.TABLE_ID)
```
Comment (cathay4t) [inline nmstatectl/nmstatectl.py]: The state does not contain `Route.TABLE_ID`.

**Comment by cathay4t** [comment]
> @cathay4t
> @ffmancera
> There is no example for route rule. Shall I work on that?

No, I created one https://github.com/nmstate/nmstate/pull/633 but closed by our bad robot due to been stale for a while. I will pick  that up.

File: nmstatectl/nmstatectl.py (hunk 6 above)
Comment (SudhanAruna) [reply nmstatectl/nmstatectl.py]: Sorry, I'll work on that. Thank you for the example yml file

**Comment by ffmancera** [comment]
I am going to review it today. Sorry for the late reply.

**Comment by SudhanAruna** [comment]
> 
> 
> Looks good to me, thanks! Please rebase :-)
I rebased branch. Thank you the review

**Comment by cathay4t** [comment]
Merged. Thanks.

**Comment by SudhanAruna** [comment]
Thamk you for reviews.

**Pull request merged**